src/include/storage/record_buffer.h

```diff
@@ -0,0 +1,189 @@
+ *
+ * Not thread-safe
+ */
+class UndoBuffer {
+ public:
+  /**
+   * Iterator to access DeltaRecords inside the undo buffer.
+   *
+   * Iterators are always guaranteed to be valid once handed out, provided that the UndoBuffer is not
+   * destructed, regardless of modifications to the UndoBuffer.
+   *
+   * Not thread-safe
+   */
+  class Iterator {
+   public:
+    /**
+     * @return reference to the underlying UndoRecord
+     */
+    UndoRecord &operator*() const {
+      return *reinterpret_cast<UndoRecord *>((*curr_segment_)->bytes_ + segment_offset_);
+    }
+
+    /**
+     * @return pointer to the underlying UndoRecord
+     */
+    UndoRecord *operator->() const {
+      return reinterpret_cast<UndoRecord *>((*curr_segment_)->bytes_ + segment_offset_);
+    }
+
+    /**
+     * prefix-increment
+     * @return self-reference
+     */
+    Iterator &operator++() {
+      UndoRecord &me = this->operator*();
+      segment_offset_ += me.Size();
+      if (segment_offset_ == (*curr_segment_)->end_) {
+        // need to advance into the next segment
+        ++curr_segment_;
+        segment_offset_ = 0;
+      }
+
+      return *this;
+    }
+
+    /**
+     * postfix-increment
+     * @return iterator equal to this iterator before increment
+     */
+    const Iterator operator++(int) {
+      Iterator copy = *this;
+      operator++();
+      return copy;
+    }
+
+    /**
+     * Equality test
+     * @param other iterator to compare to
+     * @return if this is equal to other
+     */
+    bool operator==(const Iterator &other) const {
+      return segment_offset_ == other.segment_offset_ && curr_segment_ == other.curr_segment_;
+    }
+
+    /**
+     * Inequality test
+     * @param other iterator to compare to
+     * @return if this is not equal to other
+     */
+    bool operator!=(const Iterator &other) const { return !(*this == other); }
+
+   private:
+    friend class UndoBuffer;
+    Iterator(std::vector<BufferSegment *>::iterator curr_segment, uint32_t segment_offset)
+        : curr_segment_(curr_segment), segment_offset_(segment_offset) {}
+    std::vector<BufferSegment *>::iterator curr_segment_;
+    uint32_t segment_offset_;
+  };
+
+  /**
+   * Constructs a new undo buffer, drawing its segments from the given buffer pool.
+   * @param buffer_pool buffer pool to draw segments from
+   */
+  explicit UndoBuffer(common::ObjectPool<BufferSegment> *buffer_pool) : buffer_pool_(buffer_pool) {}
+
+  /**
+   * Destructs this buffer, releases all its segments back to the buffer pool it draws from.
+   */
+  ~UndoBuffer() {
+    for (auto *segment : buffers_) buffer_pool_->Release(segment);
+  }
+
+  /**
+   * @return Iterator to the first element
+   */
+  Iterator begin() { return {buffers_.begin(), 0}; }
+
+  /**
+   * @return Iterator to the element following the last element
+   */
+  Iterator end() { return {buffers_.end(), 0}; }
+
+  /**
+   * @return true if UndoBuffer contains no DeltaRecords, false otherwise
```

Comment:
Suggest just searching the entire codebase for "DeltaRecord"

Reply:
CLion lies about refactoring the comments as well

One exception is that I might keep the DeltaRecordTests since we will probably just write RedoRecord test in there as well.